functional_programming_tests/test/with_dependency_injection.cpp

```diff
@@ -0,0 +1,528 @@
+#include <chrono>
+#include <cmath>
+#include <iostream>
+#include <memory>
+#include <optional>
+#include <vector>
+
+#include <rclcpp/rclcpp.hpp>
+
+#include <example_srvs/srv/get_path.hpp>
+#include <example_srvs/srv/set_map.hpp>
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+#include <std_msgs/msg/u_int8_multi_array.hpp>
+
+using namespace std::chrono_literals;
+
+/**
+ * @brief Represents a position in the map
+ */
+struct Position {
+  size_t x;
+  size_t y;
+};
+
+// Operator overload for position comparison
+bool operator==(Position const& lhs, Position const& rhs) {
+  return lhs.x == rhs.x && lhs.y == rhs.y;
+}
+
+using Path = std::vector<Position>;
+
+// Operator overload to print path
+std::ostream& operator<<(std::ostream& os, Path const& path) {
+  for (auto const& pose : path) {
+    os << "(" << pose.x << ", " << pose.y << ")\n";
+  }
+  return os;
+}
+
+template <typename T>
+class Map {
+ public:
+  Map(){};
+  /**
+   * @brief Construct a map from a 2D vector
+   * @param data The 2D vector to construct the map from
+   */
+  Map(std::vector<std::vector<T>> data) : data_{data} {};
+  T at(Position const& pos) const { return data_.at(pos.y).at(pos.x); }
+  std::vector<std::vector<T>>& get_data() { return data_; }
+
+ private:
+  std::vector<std::vector<T>> data_;
+};
+
+struct PathGenerator {
+  struct MiddlewareHandle {
```

Comment:
Correct. Though, to be clear:
- It is public so that the ros middleware and mock middleware can inherit from it
- It is declared in the `PathGenerator` class because I like that style to say that it is specific to this class. Using a namespace would be equally as valid (or just making a longer name)
- If a generic version would work, say by implementing all of the components to abstract and then wrapping those up in a parameter object would probably also work